File: Java-Core/src/main/java/presto/statement/SwitchStatement.java
package presto.statement;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class SwitchStatement extends BaseSwitchStatement {

	IExpression expression;
	
	public SwitchStatement(IExpression expression) {
		super();
		this.expression = expression;
	}

	public SwitchStatement(IExpression expression, SwitchCaseList switchCases, StatementList defaultCase) {
		super(switchCases, defaultCase);
		this.expression = expression;
	}

	@Override
	protected void toODialect(CodeWriter writer) {
		writer.append("switch(");
		expression.toDialect(writer);
		writer.append(") {\n");
		for(SwitchCase sc : switchCases)
			sc.caseToODialect(writer);
		if(defaultCase!=null) {
			writer.append("default:\n");
			writer.indent();
			defaultCase.toDialect(writer);
			writer.dedent();
		}
		writer.append("}\n");
	}

	@Override
	protected void toEDialect(CodeWriter writer) {
		writer.append("switch on ");
		expression.toDialect(writer);
		writer.append(":\n");
		writer.indent();
		for(SwitchCase sc : switchCases)
			sc.caseToEDialect(writer);
		if(defaultCase!=null) {
			writer.append("otherwise:\n");
			writer.indent();
			defaultCase.toDialect(writer);
			writer.dedent();
		}
		writer.dedent();
	}

	@Override
	protected void toPDialect(CodeWriter writer) {
		toEDialect(writer);
	}

	@Override
	IType checkSwitchType(Context context) throws SyntaxError {
		return expression.check(context);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		IValue switchValue = expression.interpret(context);
		return evaluateSwitch(context,switchValue,null);
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/statement/SwitchStatement.js
var BaseSwitchStatement = require("./BaseSwitchStatement").BaseSwitchStatement;

function SwitchStatement(expression, switchCases, defaultCase) {
	BaseSwitchStatement.call(this, switchCases, defaultCase);
	this.expression = expression;
}

SwitchStatement.prototype = Object.create(BaseSwitchStatement.prototype);
SwitchStatement.prototype.constructor = SwitchStatement;

SwitchStatement.prototype.checkSwitchType = function(context) {
	return this.expression.check(context);
};

SwitchStatement.prototype.interpret = function(context) {
	var switchValue = this.expression.interpret(context);
	return this.interpretSwitch(context,switchValue,null);
};

SwitchStatement.prototype.toODialect = function(writer) {
    writer.append("switch(");
    this.expression.toDialect(writer);
    writer.append(") {\n");
    for(var i=0;i<this.switchCases.length;i++)
        this.switchCases[i].caseToODialect(writer);
    if(this.defaultCase!=null) {
        writer.append("default:\n");
        writer.indent();
        this.defaultCase.toDialect(writer);
        writer.dedent();
    }
    writer.append("}\n");
}

SwitchStatement.prototype.toEDialect = function(writer) {
    writer.append("switch on ");
    this.expression.toDialect(writer);
    writer.append(":\n");
    writer.indent();
    for(var i=0;i<this.switchCases.length;i++)
        this.switchCases[i].caseToEDialect(writer);
    if(this.defaultCase!=null) {
        writer.append("otherwise:\n");
        writer.indent();
        this.defaultCase.toDialect(writer);
        writer.dedent();
    }
    writer.dedent();
}

SwitchStatement.prototype.toSDialect = function(writer) {
    writer.append("switch on ");
    this.expression.toDialect(writer);
    writer.append(":\n");
    writer.indent();
    for(var i=0;i<this.switchCases.length;i++)
        this.switchCases[i].caseToPDialect(writer);
    if(this.defaultCase!=null) {
        writer.append("otherwise:\n");
        writer.indent();
        this.defaultCase.toDialect(writer);
        writer.dedent();
    }
    writer.dedent();
};

exports.SwitchStatement = SwitchStatement;